cpp/cmake/thirdparty/CUDF_GetThrust.cmake

```diff
@@ -14,15 +14,18 @@
 # limitations under the License.
 #=============================================================================
 
-project(cudf-GoogleTest)
+function(find_and_configure_thrust VERSION)
+    CPMAddPackage(NAME Thrust
+        VERSION         ${VERSION}
+        GIT_REPOSITORY  https://github.com/NVIDIA/thrust.git
+        GIT_TAG         ${VERSION}
+        GIT_SHALLOW     TRUE
+        PATCH_COMMAND   patch -p1 -N < ${CUDF_SOURCE_DIR}/cmake/thrust.patch || true)
 
-include(ExternalProject)
+    thrust_create_target(cudf::Thrust FROM_OPTIONS)
+    set(THRUST_LIBRARY "cudf::Thrust" PARENT_SCOPE)
+endfunction()
 
-ExternalProject_Add(GoogleTest
-                    GIT_REPOSITORY    https://github.com/google/googletest.git
-                    GIT_TAG           release-1.8.0
-                    GIT_SHALLOW       true
-                    SOURCE_DIR        "${GTEST_ROOT}/googletest"
-                    BINARY_DIR        "${GTEST_ROOT}/build"
-                    INSTALL_DIR       "${GTEST_ROOT}/install"
-                    CMAKE_ARGS        ${GTEST_CMAKE_ARGS} -DCMAKE_INSTALL_PREFIX=${GTEST_ROOT}/install)
+set(CUDF_MIN_VERSION_Thrust 1.10.0)
```

Comment:
Is this still a min version or just the version since we always pull Thrust now?

Reply:
Yeah this is really just the Thrust version. It looks like we do include Thrust in our public headers, but I'm not sure what happens if the install script finds and uses another version of Thrust on the installed system.

Reply:
Finding a different version of Thrust for headers I imagine should be okay as leaking thrust in public headers should hopefully be minimal.

Reply:
It unfortunately is not:
```shell
$ find cpp/include/cudf -type f -print0 | xargs -0 grep -n '#include <thrust'
cpp/include/cudf/column/column_device_view.cuh:33:#include <thrust/iterator/counting_iterator.h>
cpp/include/cudf/column/column_device_view.cuh:34:#include <thrust/iterator/transform_iterator.h>
cpp/include/cudf/table/row_operators.cuh:27:#include <thrust/equal.h>
cpp/include/cudf/table/row_operators.cuh:28:#include <thrust/swap.h>
cpp/include/cudf/table/row_operators.cuh:29:#include <thrust/transform_reduce.h>
cpp/include/cudf/detail/reduction_operators.cuh:24:#include <thrust/functional.h>
cpp/include/cudf/detail/valid_if.cuh:28:#include <thrust/device_vector.h>
cpp/include/cudf/detail/valid_if.cuh:29:#include <thrust/distance.h>
cpp/include/cudf/detail/indexalator.cuh:23:#include <thrust/iterator/constant_iterator.h>
cpp/include/cudf/detail/utilities/trie.cuh:31:#include <thrust/host_vector.h>
cpp/include/cudf/detail/utilities/transform_unary_functions.cuh:29:#include <thrust/iterator/transform_iterator.h>
cpp/include/cudf/detail/utilities/transform_unary_functions.cuh:30:#include <thrust/pair.h>
cpp/include/cudf/detail/gather.cuh:43:#include <thrust/functional.h>
cpp/include/cudf/detail/gather.cuh:44:#include <thrust/gather.h>
cpp/include/cudf/detail/gather.cuh:45:#include <thrust/host_vector.h>
cpp/include/cudf/detail/gather.cuh:46:#include <thrust/iterator/transform_iterator.h>
cpp/include/cudf/detail/gather.cuh:47:#include <thrust/logical.h>
cpp/include/cudf/detail/reduction.cuh:30:#include <thrust/for_each.h>
cpp/include/cudf/detail/reduction.cuh:31:#include <thrust/iterator/iterator_traits.h>
cpp/include/cudf/types.hpp:27:#include <thrust/optional.h>  // TODO no idea why this is needed ¯\_(ツ)_/¯
cpp/include/cudf/utilities/span.hpp:23:#include <thrust/detail/raw_pointer_cast.h>
cpp/include/cudf/utilities/span.hpp:24:#include <thrust/device_vector.h>
cpp/include/cudf/lists/detail/gather.cuh:26:#include <thrust/transform_scan.h>
cpp/include/cudf/lists/detail/scatter.cuh:33:#include <thrust/binary_search.h>
cpp/include/cudf/strings/string_view.inl:17:#include <thrust/find.h>
cpp/include/cudf/strings/detail/sorting.cuh:27:#include <thrust/sequence.h>
cpp/include/cudf/strings/detail/sorting.cuh:28:#include <thrust/sort.h>
cpp/include/cudf/strings/detail/strings_column_factories.cuh:29:#include <thrust/for_each.h>
cpp/include/cudf/strings/detail/strings_column_factories.cuh:30:#include <thrust/transform_reduce.h>
cpp/include/cudf/strings/detail/utilities.cuh:24:#include <thrust/scan.h>
cpp/include/cudf/strings/detail/copy_range.cuh:29:#include <thrust/for_each.h>
cpp/include/cudf/strings/detail/copy_range.cuh:30:#include <thrust/iterator/counting_iterator.h>
cpp/include/cudf/strings/detail/copy_range.cuh:31:#include <thrust/iterator/transform_iterator.h>
cpp/include/cudf/strings/string.cuh:18:#include <thrust/logical.h>
```

Reply:
Oh boy... I guess for now we just specify a minimum version and hope for the best?

Should we consider installing thrust inside the cudf folder in some way and then using those headers? I.E. something like `#include <cudf/thrust/logical.h>`? When someone installs libcudf we also install our own copy of thrust inside of it?

Reply:
We do that for `libcudacxx`, but I'm not sure whether it's appropriate to do it for thrust. @harrism @jrhemstad thoughts?

Reply:
Nearly all of those things are `detail` headers, so I'm not sure what the problem is.

Reply:
I think the case we're concerned about here is when someone goes to build something against libcudf it won't have our specific version and patches to Thrust and whether that is an issue or not.